builder/upcloud/utils.go

```diff
@@ -0,0 +1,32 @@
+package upcloud
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/UpCloudLtd/upcloud-go-api/upcloud"
+	"github.com/hashicorp/packer-plugin-sdk/multistep"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
+)
+
+// wraps error logic
+func StepHaltWithError(state multistep.StateBag, err error) multistep.StepAction {
+	ui := state.Get("ui").(packersdk.Ui)
```

Comment:
Not familiar with Packer plugin SDK, so this might indeed be the correct way, but aren't there really any structures for holding this information? Type converting a lot of things from here with just hoping they'll be there seems kinda strange. :o

Reply:
I've double-checked and using StateBug to pass variables is common practice for packer-plugin-sdk. Probably we could design something our own for that purpose, but tbh I'll preferer to leave it as it is. Would be easier to maintain and upgrade to newer versions of sdk.